File: app/src/main/java/cu/axel/smartdock/utils/AppUtils.java
package cu.axel.smartdock.utils;

import android.app.ActivityManager;
import android.app.usage.UsageStats;
import android.app.usage.UsageStatsManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.SystemClock;
import android.util.Log;

import androidx.preference.PreferenceManager;
import cu.axel.smartdock.models.DockApp;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import cu.axel.smartdock.models.App;
import cu.axel.smartdock.models.AppTask;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class AppUtils {
	public static final String PINNED_LIST = "pinned.lst";
	public static final String DOCK_PINNED_LIST = "dock_pinned.lst";
	public static final String DESKTOP_LIST = "desktop.lst";
	public static String currentApp = "";

	public enum WindowMode {
		STANDARD, PORTRAIT, MAXIMIZED, FULLSCREEN, TILED_TOP, TILED_LEFT, TILED_RIGHT, TILED_BOTTOM
	}

	public static ArrayList<App> getInstalledApps(PackageManager pm) {
		ArrayList<App> apps = new ArrayList<>();
		Intent intent = new Intent(Intent.ACTION_MAIN);
		intent.addCategory(Intent.CATEGORY_LAUNCHER);
		List<ResolveInfo> appsInfo = pm.queryIntentActivities(intent, 0);

		//TODO: Filter Google App
		for (ResolveInfo appInfo : appsInfo) {
			String label = appInfo.activityInfo.loadLabel(pm).toString();
			Drawable icon = appInfo.activityInfo.loadIcon(pm);
			String packageName = appInfo.activityInfo.packageName;

			apps.add(new App(label, packageName, icon));
		}

		Collections.sort(apps, (p1, p2) -> p1.getName().compareToIgnoreCase(p2.getName()));

		return apps;
	}

	public static ArrayList<App> getPinnedApps(Context context, PackageManager pm, String type) {
		ArrayList<App> apps = new ArrayList<>();

		try {
			BufferedReader br = new BufferedReader(new FileReader(new File(context.getFilesDir(), type)));
			String applist;
			try {
				if ((applist = br.readLine()) != null) {
					String[] applist2 = applist.split(" ");
					for (String app : applist2) {
						try {
							ApplicationInfo appInfo = pm.getApplicationInfo(app, 0);
							apps.add(new App(pm.getApplicationLabel(appInfo).toString(), app,
									pm.getApplicationIcon(app)));
						} catch (PackageManager.NameNotFoundException e) {
							//app is no longer available, lets unpin it
							unpinApp(context, app, type);
						}
					}
				}

			} catch (IOException e) {
			}
		} catch (FileNotFoundException e) {
		}

		return apps;

	}

	public static void pinApp(Context context, String app, String type) {
		try {
			File file = new File(context.getFilesDir(), type);
			FileWriter fw = new FileWriter(file, true);
			fw.write(app + " ");
			fw.close();
		} catch (IOException e) {
		}

	}

	public static void unpinApp(Context context, String app, String type) {
		try {
			File file = new File(context.getFilesDir(), type);
			BufferedReader br = new BufferedReader(new FileReader(file));
			String applist;

			if ((applist = br.readLine()) != null) {
				applist = applist.replace(app + " ", "");
				FileWriter fw = new FileWriter(file, false);
				fw.write(applist);
				fw.close();
			}

		} catch (IOException e) {
		}
	}

	public static void moveApp(Context context, String app, String type, int direction) {
		try {
			File file = new File(context.getFilesDir(), type);
			BufferedReader br = new BufferedReader(new FileReader(file));
			String applist, what = "", with = "";

			if ((applist = br.readLine()) != null) {
				String[] apps = applist.split(" ");
				int pos = findInArray(app, apps);
				if (direction == 0 && pos > 0) {
					what = apps[pos - 1] + " " + app;
					with = app + " " + apps[pos - 1];
				} else if (direction == 1 && pos < apps.length - 1) {
					what = app + " " + apps[pos + 1];
					with = apps[pos + 1] + " " + app;
				}
				applist = applist.replace(what, with);
				FileWriter fw = new FileWriter(file, false);
				fw.write(applist);
				fw.close();
			}

		} catch (IOException e) {
		}
	}

	public static int findInArray(String key, String[] array) {
		for (int i = 0; i < array.length; i++) {
			if (array[i].contains(key))
				return i;
		}
		return -1;
	}

	public static boolean isPinned(Context context, String app, String type) {
		try {
			BufferedReader br = new BufferedReader(new FileReader(new File(context.getFilesDir(), type)));
			String applist;

			if ((applist = br.readLine()) != null) {
				return applist.contains(app);
			}

		} catch (IOException e) {
		}
		return false;
	}

	public static boolean isGame(PackageManager pm, String packageName) {
		try {
			ApplicationInfo info = pm.getApplicationInfo(packageName, 0);
			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
				return info.category == ApplicationInfo.CATEGORY_GAME;
			} else {
				return (info.flags & ApplicationInfo.FLAG_IS_GAME) == ApplicationInfo.FLAG_IS_GAME;
			}
		} catch (PackageManager.NameNotFoundException e) {
			return false;
		}
	}

	public static String getCurrentLauncher(PackageManager pm) {
		Intent intent = new Intent(Intent.ACTION_MAIN);
		intent.addCategory(Intent.CATEGORY_HOME);
		ResolveInfo resolveInfo = pm.resolveActivity(intent, PackageManager.MATCH_DEFAULT_ONLY);
		return resolveInfo.activityInfo.packageName;
	}

	public static void setWindowMode(ActivityManager am, int taskId, int mode) {
		try {
			Method setWindowMode = am.getClass().getMethod("setTaskWindowingMode", int.class, int.class, boolean.class);
			setWindowMode.invoke(am, taskId, mode, false);
		} catch (Exception e) {
		}
	}

	public static ArrayList<AppTask> getRunningTasks(ActivityManager am, PackageManager pm, int max) {
		List<ActivityManager.RunningTaskInfo> tasksInfo = am.getRunningTasks(max);
		currentApp = tasksInfo.get(0).baseActivity.getPackageName();

		ArrayList<AppTask> appTasks = new ArrayList<>();
		for (ActivityManager.RunningTaskInfo taskInfo : tasksInfo) {
			try {
				//Exclude systemui, launcher and other system apps from the tasklist
				if (taskInfo.baseActivity.getPackageName().contains("com.android.systemui")
						|| taskInfo.baseActivity.getPackageName().contains("com.google.android.packageinstaller"))
					continue;

				//Hack to save Dock settings activity ftom being excluded
				if (!(taskInfo.topActivity.getClassName().equals("cu.axel.smartdock.activities.MainActivity")
						|| taskInfo.topActivity.getClassName().equals("cu.axel.smartdock.activities.DebugActivity"))
						&& taskInfo.topActivity.getPackageName().equals(AppUtils.getCurrentLauncher(pm)))
					continue;

				if (Build.VERSION.SDK_INT > 29) {
					try {
						Field isRunning = taskInfo.getClass().getField("isRunning");
						boolean running = isRunning.getBoolean(taskInfo);
						if (!running)
							continue;
					} catch (Exception e) {
					}
				}

				appTasks.add(
						new AppTask(taskInfo.id, pm.getActivityInfo(taskInfo.topActivity, 0).loadLabel(pm).toString(),
								taskInfo.topActivity.getPackageName(), pm.getActivityIcon(taskInfo.topActivity)));
			} catch (PackageManager.NameNotFoundException e) {
			}
		}

		return appTasks;
	}

	public static ArrayList<AppTask> getRecentTasks(Context context, int max) {
		UsageStatsManager usm = (UsageStatsManager) context.getSystemService(Context.USAGE_STATS_SERVICE);
		long start = System.currentTimeMillis() - SystemClock.elapsedRealtime();
		List<UsageStats> usageStats = usm.queryUsageStats(UsageStatsManager.INTERVAL_BEST, start,
				System.currentTimeMillis());
		ArrayList<AppTask> appTasks = new ArrayList<>();

		Collections.sort(usageStats, (p1, p2) -> Long.compare(p2.getLastTimeUsed(), p1.getLastTimeUsed()));

		for (UsageStats stat : usageStats) {
			String app = stat.getPackageName();
			try {
				if (isLaunchable(context, app) && !app.equals(AppUtils.getCurrentLauncher(context.getPackageManager())))
					appTasks.add(new AppTask(-1, getPackageLabel(context, app), app,
							context.getPackageManager().getApplicationIcon(app)));
			} catch (PackageManager.NameNotFoundException e) {
			}

			if (appTasks.size() >= max)
				break;
		}

		return appTasks;
	}

	public static boolean isSystemApp(Context context, String app) {
		try {
			ApplicationInfo appInfo = context.getPackageManager().getApplicationInfo(app, 0);
			return (appInfo.flags & (ApplicationInfo.FLAG_SYSTEM | ApplicationInfo.FLAG_UPDATED_SYSTEM_APP)) != 0;
		} catch (PackageManager.NameNotFoundException e) {
			return false;
		}
	}

	private static boolean isLaunchable(Context context, String app) {
		List<ResolveInfo> resolveInfo = context.getPackageManager().queryIntentActivities(
				new Intent(Intent.ACTION_MAIN).addCategory(Intent.CATEGORY_LAUNCHER).setPackage(app), 0);
		return resolveInfo.size() > 0;
	}

	public static void removeTask(ActivityManager am, int id) {
		try {
			Method removeTask = am.getClass().getMethod("removeTask", int.class);
			removeTask.invoke(am, id);
		} catch (Exception e) {
			Log.e("Dock", e + e.getCause().toString());
		}
	}

	public static String getPackageLabel(Context context, String packageName) {
		try {
			PackageManager pm = context.getPackageManager();
			ApplicationInfo appInfo = pm.getApplicationInfo(packageName, 0);
			return pm.getApplicationLabel(appInfo).toString();
		} catch (PackageManager.NameNotFoundException e) {
		}
		return "";
	}

	public static Drawable getAppIcon(Context context, String app) {
		try {
			return context.getPackageManager().getApplicationIcon(app);
		} catch (PackageManager.NameNotFoundException e) {
		}
		return context.getDrawable(android.R.drawable.sym_def_app_icon);
	}

	public static Rect makeLaunchBounds(Context context, String mode, int dockHeight, boolean secondary) {

		SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(context);
		int left = 0, top = 0, right = 0, bottom = 0;
		int deviceWidth = DeviceUtils.getDisplayMetrics(context, secondary).widthPixels;
		int deviceHeight = DeviceUtils.getDisplayMetrics(context, secondary).heightPixels;
		int statusHeight = DeviceUtils.getStatusBarHeight(context);
		int usableHeight = deviceHeight - dockHeight - statusHeight;
		float scaleFactor = Float.parseFloat(sp.getString("scale_factor", "1.0"));

		switch (mode) {
		case "standard":
			left = (int) (deviceWidth / (5 * scaleFactor));
			top = (int) ((usableHeight + statusHeight) / (7 * scaleFactor));
			right = deviceWidth - left;
			bottom = usableHeight + dockHeight - top;
			break;
		case "maximized":
			right = deviceWidth;
			bottom = usableHeight;
			break;
		case "portrait":
			left = deviceWidth / 3;
			top = (usableHeight) / 15;
			right = deviceWidth - left;
			bottom = usableHeight + dockHeight - top;
			break;
		case "tiled-left":
			right = deviceWidth / 2;
			bottom = usableHeight;
			break;
		case "tiled-top":
			right = deviceWidth;
			bottom = (usableHeight + statusHeight) / 2;
			break;
		case "tiled-right":
			left = deviceWidth / 2;
			right = deviceWidth;
			bottom = usableHeight;
			break;
		case "tiled-bottom":
			right = deviceWidth;
			top = (usableHeight + statusHeight) / 2;
			bottom = usableHeight + statusHeight;
			break;
		}

		Rect launchBounds = new Rect(left, top, right, bottom);

		return launchBounds;
	}

	public static void resizeTask(Context context, String mode, int taskId, int dockHeight, boolean secondary) {
		if (taskId < 0)
			return;
		Rect bounds = makeLaunchBounds(context, mode, dockHeight, secondary);
		DeviceUtils.runAsRoot("am task resize " + taskId + " " + bounds.left + " " + bounds.top + " " + bounds.right
				+ " " + bounds.bottom);
	}

	public static int containsTask(ArrayList<DockApp> apps, AppTask task) {
		for (int i = 0; i < apps.size(); i++) {
			if (apps.get(i).getPackageName().equals(task.getPackageName()))
				return i;
		}
		return -1;
	}

	public static String findStackId(Context context, int taskId) {
		String stackInfo = DeviceUtils.runAsRoot("am stack list");
		String regexPattern = "(?s)Stack id=(\\d+).*?taskId=" + taskId;
		Pattern pattern = Pattern.compile(regexPattern);
		Matcher matcher = pattern.matcher(stackInfo);

		if (matcher.find())
			return matcher.group(1);

		return null;
	}

	public static void removeStack(String stackId) {
		DeviceUtils.runAsRoot("am stack remove " + stackId);
	}

}

File: app/src/main/java/cu/axel/smartdock/utils/AppUtils.kt
package cu.axel.smartdock.utils

import android.app.ActivityManager
import android.app.ActivityOptions
import android.app.usage.UsageStats
import android.app.usage.UsageStatsManager
import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.LauncherActivityInfo
import android.content.pm.LauncherApps
import android.content.pm.PackageManager
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.SystemClock
import android.os.UserManager
import android.view.Display
import androidx.appcompat.content.res.AppCompatResources
import androidx.preference.PreferenceManager
import cu.axel.smartdock.models.App
import cu.axel.smartdock.models.AppTask
import cu.axel.smartdock.models.DockApp
import java.io.File

object AppUtils {
    const val PINNED_LIST = "pinned.lst"
    const val DOCK_PINNED_LIST = "dock_pinned.lst"
    const val DESKTOP_LIST = "desktop.lst"
    var currentApp = ""
    fun getInstalledApps(context: Context): ArrayList<App> {
        val apps = ArrayList<App>()
        val userManager = context.getSystemService(Context.USER_SERVICE) as UserManager
        val launcherApps = context.getSystemService(Context.LAUNCHER_APPS_SERVICE) as LauncherApps
        var appsInfo = mutableListOf<LauncherActivityInfo>()
        for (profile in userManager.userProfiles)
            appsInfo.addAll(launcherApps.getActivityList(null, profile))

        appsInfo = appsInfo.sortedWith(compareBy { it.label.toString() }).toMutableList()


        //TODO: Filter Google App
        for (appInfo in appsInfo) {
            apps.add(
                App(
                    appInfo.label.toString(), appInfo.componentName.packageName, appInfo.getIcon(0),
                    appInfo.componentName, appInfo.user
                )
            )
        }

        return apps
    }

    fun getPinnedApps(context: Context, type: String): ArrayList<App> {
        val file = File(context.filesDir, type)
        val apps = ArrayList<App>()
        val appsInfo = mutableListOf<LauncherActivityInfo>()
        val launcherApps = context.getSystemService(Context.LAUNCHER_APPS_SERVICE) as LauncherApps
        val userManager = context.getSystemService(Context.USER_SERVICE) as UserManager
        if (file.exists()) {
            for (line in file.readLines()) {
                if (line.isBlank())
                    continue
                val info = line.split(" ")
                val packageName = info[0]
                val userHandle = userManager.getUserForSerialNumber(info[1].toLong())
                val list = launcherApps.getActivityList(packageName, userHandle)
                if (list.isNullOrEmpty())
                    unpinApp(context, packageName, type)
                appsInfo.addAll(list)
            }
        }

        for (appInfo in appsInfo) {
            apps.add(
                App(
                    appInfo.label.toString(), appInfo.componentName.packageName, appInfo.getIcon(0),
                    appInfo.componentName, appInfo.user
                )
            )
        }

        return apps
    }

    fun pinApp(context: Context, app: App, type: String) {
        val userManager = context.getSystemService(Context.USER_SERVICE) as UserManager
        val file = File(context.filesDir, type)
        file.appendText("${app.packageName} ${userManager.getSerialNumberForUser(app.userHandle)}\n")
    }

    fun unpinApp(context: Context, packageName: String, type: String) {
        val file = File(context.filesDir, type)
        val updatedList = file.readLines().filter { it.split(" ")[0] != packageName }
        if (updatedList.isNotEmpty())
            file.writeText(updatedList.joinToString("\n") + "\n")
        else {
            file.writeText("")
        }
    }

    fun moveApp(context: Context, app: App, type: String, direction: Int) {
        val file = File(context.filesDir, type)
        val lines = file.readLines().toMutableList()

        val lineIndex = lines.indexOfFirst { it.split(" ")[0] == app.packageName }

        if (lineIndex != -1) {
            if (direction == 0 && lineIndex > 0) {
                val line = lines.removeAt(lineIndex)
                lines.add(lineIndex - 1, line)
            } else if (direction == 1 && lineIndex < lines.size - 1) {
                val line = lines.removeAt(lineIndex)
                lines.add(lineIndex + 1, line)
            }

            file.writeText(lines.joinToString("\n") + "\n")
        }
    }

    fun isPinned(context: Context, app: App, type: String): Boolean {
        val file = File(context.filesDir, type)
        if (!file.exists())
            return false
        file.readLines().forEach { line ->
            if (line.split(" ")[0] == app.packageName)
                return true
        }
        return false
    }

    fun isGame(packageManager: PackageManager, packageName: String): Boolean {
        return try {
            val info = packageManager.getApplicationInfo(packageName, 0)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                info.category == ApplicationInfo.CATEGORY_GAME
            } else {
                info.flags and ApplicationInfo.FLAG_IS_GAME == ApplicationInfo.FLAG_IS_GAME
            }
        } catch (e: PackageManager.NameNotFoundException) {
            false
        }
    }

    private fun getCurrentLauncher(packageManager: PackageManager): String {
        val intent = Intent(Intent.ACTION_MAIN)
        intent.addCategory(Intent.CATEGORY_HOME)
        val resolveInfo = packageManager.resolveActivity(intent, PackageManager.MATCH_DEFAULT_ONLY)
        return resolveInfo!!.activityInfo.packageName
    }

    fun setWindowMode(activityManager: ActivityManager, taskId: Int, mode: Int) {
        try {
            val setWindowMode = activityManager.javaClass.getMethod(
                "setTaskWindowingMode",
                Int::class.javaPrimitiveType,
                Int::class.javaPrimitiveType,
                Boolean::class.javaPrimitiveType
            )
            setWindowMode.invoke(activityManager, taskId, mode, false)
        } catch (_: Exception) {
        }
    }

    fun getRunningTasks(
        activityManager: ActivityManager,
        packageManager: PackageManager,
        max: Int
    ): ArrayList<AppTask> {
        val tasksInfo = activityManager.getRunningTasks(max)
        currentApp = tasksInfo[0].baseActivity!!.packageName
        val appTasks = ArrayList<AppTask>()
        for (taskInfo in tasksInfo) {
            try {
                //Exclude systemui, launcher and other system apps from the tasklist
                if (taskInfo.baseActivity!!.packageName.contains("com.android.systemui")
                    || taskInfo.baseActivity!!.packageName.contains("com.google.android.packageinstaller")
                    || taskInfo.baseActivity!!.className == "com.android.quickstep.RecentsActivity"
                ) continue

                //Hack to save Dock settings activity ftom being excluded
                if (!(taskInfo.topActivity!!.className == "cu.axel.smartdock.activities.MainActivity" || taskInfo.topActivity!!.className == "cu.axel.smartdock.activities.DebugActivity") && taskInfo.topActivity!!.packageName == getCurrentLauncher(
                        packageManager
                    )
                ) continue
                if (Build.VERSION.SDK_INT > 29) {
                    try {
                        val isRunning = taskInfo.javaClass.getField("isRunning")
                        val running = isRunning.getBoolean(taskInfo)
                        if (!running) continue
                    } catch (_: Exception) {
                    }
                }
                appTasks.add(
                    AppTask(
                        taskInfo.id,
                        packageManager.getActivityInfo(taskInfo.topActivity!!, 0)
                            .loadLabel(packageManager).toString(),
                        taskInfo.topActivity!!.packageName,
                        packageManager.getActivityIcon(taskInfo.topActivity!!)
                    )
                )
            } catch (_: PackageManager.NameNotFoundException) {
            }
        }
        return appTasks
    }

    fun getRecentTasks(context: Context, max: Int): ArrayList<AppTask> {
        val usm = context.getSystemService(Context.USAGE_STATS_SERVICE) as UsageStatsManager
        val start = System.currentTimeMillis() - SystemClock.elapsedRealtime()
        val usageStats = usm.queryUsageStats(
            UsageStatsManager.INTERVAL_BEST, start,
            System.currentTimeMillis()
        )
        val appTasks = ArrayList<AppTask>()
        usageStats.sortWith { usageStats1: UsageStats, usageStats2: UsageStats ->
            usageStats2.lastTimeUsed.compareTo(
                usageStats1.lastTimeUsed
            )
        }
        for (stat in usageStats) {
            val app = stat.packageName
            try {
                if (isLaunchable(
                        context,
                        app
                    ) && app != getCurrentLauncher(context.packageManager)
                ) appTasks.add(
                    AppTask(
                        -1, getPackageLabel(context, app), app,
                        context.packageManager.getApplicationIcon(app)
                    )
                )
            } catch (_: PackageManager.NameNotFoundException) {
            }
            if (appTasks.size >= max) break
        }
        return appTasks
    }

    fun isSystemApp(context: Context, app: String): Boolean {
        return try {
            val appInfo = context.packageManager.getApplicationInfo(app, 0)
            appInfo.flags and (ApplicationInfo.FLAG_SYSTEM or ApplicationInfo.FLAG_UPDATED_SYSTEM_APP) != 0
        } catch (e: PackageManager.NameNotFoundException) {
            false
        }
    }

    private fun isLaunchable(context: Context, app: String): Boolean {
        val resolveInfo = context.packageManager.queryIntentActivities(
            Intent(Intent.ACTION_MAIN).addCategory(Intent.CATEGORY_LAUNCHER).setPackage(app), 0
        )
        return resolveInfo.size > 0
    }

    fun getPackageLabel(context: Context, packageName: String): String {
        try {
            val packageManager = context.packageManager
            val appInfo = packageManager.getApplicationInfo(packageName, 0)
            return packageManager.getApplicationLabel(appInfo).toString()
        } catch (_: PackageManager.NameNotFoundException) {
        }
        return ""
    }

    fun getAppIcon(context: Context, app: String): Drawable {
        return try {
            context.packageManager.getApplicationIcon(app)
        } catch (_: PackageManager.NameNotFoundException) {
            AppCompatResources.getDrawable(context, android.R.drawable.sym_def_app_icon)!!
        }
    }

    private fun makeLaunchBounds(
        context: Context,
        mode: String,
        dockHeight: Int,
        displayId: Int = Display.DEFAULT_DISPLAY
    ): Rect {
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        var left = 0
        var top = 0
        var right = 0
        var bottom = 0
        val deviceWidth = DeviceUtils.getDisplayMetrics(context, displayId).widthPixels
        val deviceHeight = DeviceUtils.getDisplayMetrics(context, displayId).heightPixels
        val statusHeight = DeviceUtils.getStatusBarHeight(context)
        val navHeight = DeviceUtils.getNavBarHeight(context)
        val diff = if (dockHeight - navHeight > 0) dockHeight - navHeight else 0

        val usableHeight =
            if (DeviceUtils.shouldApplyNavbarFix())
                deviceHeight - diff - DeviceUtils.getStatusBarHeight(context)
            else
                deviceHeight - dockHeight - DeviceUtils.getStatusBarHeight(context)
        val scaleFactor = sharedPreferences.getString("scale_factor", "1.0")!!.toFloat()
        when (mode) {
            "standard" -> {
                left = (deviceWidth / (5 * scaleFactor)).toInt()
                top = ((usableHeight + statusHeight) / (7 * scaleFactor)).toInt()
                right = deviceWidth - left
                bottom = usableHeight + dockHeight - top
            }

            "maximized" -> {
                right = deviceWidth
                bottom = usableHeight
            }

            "portrait" -> {
                left = deviceWidth / 3
                top = usableHeight / 15
                right = deviceWidth - left
                bottom = usableHeight + dockHeight - top
            }

            "tiled-left" -> {
                right = deviceWidth / 2
                bottom = usableHeight
            }

            "tiled-top" -> {
                right = deviceWidth
                bottom = (usableHeight + statusHeight) / 2
            }

            "tiled-right" -> {
                left = deviceWidth / 2
                right = deviceWidth
                bottom = usableHeight
            }

            "tiled-bottom" -> {
                right = deviceWidth
                top = (usableHeight + statusHeight) / 2
                bottom = usableHeight + statusHeight
            }
        }
        return Rect(left, top, right, bottom)
    }

    fun makeActivityOptions(
        context: Context,
        mode: String,
        dockHeight: Int,
        displayId: Int
    ): ActivityOptions {
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        val secondary = sharedPreferences.getBoolean("prefer_last_display", false)

        val display: Int =
            if (displayId != Display.DEFAULT_DISPLAY) displayId else (if (secondary) DeviceUtils.getSecondaryDisplay(
                context
            ).displayId else displayId)
        val options: ActivityOptions = ActivityOptions.makeBasic()

        val windowMode: Int
        if (mode == "fullscreen")
            windowMode = 1
        else {
            windowMode = if (Build.VERSION.SDK_INT >= 28) 5 else 2
            options.setLaunchBounds(
                makeLaunchBounds(
                    context,
                    mode,
                    dockHeight,
                    display
                )
            )
        }
        if (Build.VERSION.SDK_INT > 28)
            options.setLaunchDisplayId(display)
        val methodName =
            if (Build.VERSION.SDK_INT >= 28) "setLaunchWindowingMode" else "setLaunchStackId"
        val method =
            ActivityOptions::class.java.getMethod(methodName, Int::class.javaPrimitiveType)
        method.invoke(options, windowMode)

        return options
    }

    fun resizeTask(context: Context, mode: String, taskId: Int, dockHeight: Int) {
        if (taskId < 0)
            return
        val bounds = makeLaunchBounds(context, mode, dockHeight)
        DeviceUtils.runAsRoot(
            "am task resize " + taskId + " " + bounds.left + " " + bounds.top + " " + bounds.right
                    + " " + bounds.bottom
        )
    }

    fun containsTask(apps: ArrayList<DockApp>, task: AppTask): Int {
        for (i in apps.indices) {
            if (apps[i].packageName == task.packageName) return i
        }
        return -1
    }

}
